### PR TITLE
hv_tlbflush.py: Evaluate the execution time with flags "hv_tlbflush"

### PR DESCRIPTION
**Pull request by kimi1978wy**
Using hv_tlbflush_code.exe tool to evaluate time.

id: 1851687

Signed-off-by: Yu Wang <wyu@redhat.com>

### CODE REVIEW

**Comment by kimi1978wy** [comment]
(1/3) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.i386.io-github-autotest-qemu.hv_tlbflush.q35: PASS (897.92 s)
 (2/3) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.hv_tlbflush.q35: PASS (1000.08 s)
 (3/3) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.hv_tlbflush.q35: PASS (667.73 s)

@YongxueHong  Could you help to review it? Thanks a lot!

File: qemu/tests/hv_tlbflush.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        time_with_flag = _run_tlbflush(session)
+
+        error_context.context("Compare test results between 2 tests")
+        factor = time_with_flag / time_without_flag
+        is_pass = True
+        if params.get("vm_arch_name", "") == "x86_64":
+            is_pass = (factor < 0.5)
+        else:
+            is_pass = (factor < 1.0)
+        if not is_pass:
```
Comment (YongxueHong) [inline qemu/tests/hv_tlbflush.py:182]: How about?
`if factor >= 0.5 if params.get("vm_arch_name") == "x86_64" else factor >= 1.0:`

File: qemu/tests/cfg/hv_tlbflush.cfg [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    stress_setup_cmd += "&& tar xvf stress-1.0.4.tar.gz && cd stress-1.0.4 && ./configure && make"
+    run_stress_cmd = "cd %s/stress-1.0.4/src && ./${stress_name} --cpu %s > /dev/null 2>&1&"
+    kill_stress_cmd = "pkill stress"
+
+    test_file = "c:\tlb_flush_test.txt"
+    test_file_size = 1G
+    delete_test_file_cmd = "del /f ${test_file}"
+    create_test_file_cmd = "${delete_test_file_cmd} && fsutil file createnew ${test_file} %s"
+
+    run_tlbflush_cmd = 'powershell -command "$m=0; for ($i=0; $i -lt 30; $i++) { $s = Get-Date; c:\${executable_name} ${test_file}; $e=Get-Date; $t = $e - $s; $m=$m+$t; Write-Host $t }; Write-Host $m"'
```
Comment (YongxueHong) [inline qemu/tests/cfg/hv_tlbflush.cfg:31]: `run_tlbflush_cmd` is too long, please seperate it into multi-lines.

File: qemu/tests/hv_tlbflush.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            test_file_size, order_magnitude="B")
+        session.cmd(create_test_file_cmd % test_file_size)
+        vm.graceful_shutdown(timeout=timeout)
+
+        logging.info("Copy & compile stress tool on host")
+        stress_setup_cmd = params["stress_setup_cmd"]
+        stress_name = params.get("stress_name", "stress")
+        stress_path = data_dir.get_deps_dir(stress_name)
+        stress_setup_cmd %= (tmp_dir, stress_path)
+        process.system(stress_setup_cmd, shell=True)
```
[the remark this answers is not in the document]
Comment (kimi1978wy) [reply qemu/tests/hv_tlbflush.py:64]: utils_test.VMStress is used for guest stress, then I tried to use HOSTStress, but it seems not work, and cannot find anyone use it.  HOSTStress always occurred error for cannot get stress process, and the host stress command is "nohup stress  > /dev/null &", not stress --cpu xx.

File: qemu/tests/hv_tlbflush.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+        param session: a loggined session to send commands
+
+        return: the time result of the running, a float value
+        """
+        run_tlbflush_cmd = params["run_tlbflush_cmd"]
+        run_tlbflush_timeout = int(params.get("run_tlbflush_timeout", 3600))
+
+        logging.info("Start stress on host")
+        _start_host_stress()
```
[the remark this answers is not in the document]
Comment (kimi1978wy) [reply qemu/tests/hv_tlbflush.py:130]: since I cannot use `utils_test.VMStress`, how about using waitfor to complete this function?

**Comment by kimi1978wy** [comment]
@YongxueHong Could you help to review it?

 (1/3) Host_RHEL.m8.u3.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.i386.io-github-autotest-qemu.hv_tlbflush.q35: /nohup: redirecting stderr to stdout
\nohup: redirecting stderr to stdout
PASS (908.44 s)
 (2/3) Host_RHEL.m8.u3.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.hv_tlbflush.q35: -nohup: redirecting stderr to stdout
\nohup: redirecting stderr to stdout
PASS (760.33 s)
 (3/3) Host_RHEL.m8.u3.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.hv_tlbflush.q35: -nohup: redirecting stderr to stdout
-nohup: redirecting stderr to stdout
PASS (761.38 s)

It will prompt "-nohup: redirecting stderr to stdout \nohup: redirecting stderr to stdout" for this case, it seems the avocado-vt problem, not this script. Could you help to check it, thanks a lot!

File: qemu/tests/hv_tlbflush.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    def _run_tlbflush(session, host_stress):
+        """
+        Start running the hv_tlvflush tool on guest.
+
+        param session: a loggined session to send commands
+
+        return: the time result of the running, a float value
+        """
+        run_tlbflush_cmd = params["run_tlbflush_cmd"]
+        run_tlbflush_timeout = int(params.get("run_tlbflush_timeout", 3600))
```
Comment (YongxueHong) [inline qemu/tests/hv_tlbflush.py:145]: Please call params.get_numeric("run_tlbflush_timeout", 3600))

File: qemu/tests/hv_tlbflush.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        time_str = o.strip().split('\n')[-1]
+        time_str = time_str.split(".")[0]
+        s_t = time.strptime(time_str, "%H:%M:%S")
+        total_time = datetime.timedelta(hours=s_t.tm_hour,
+                                        minutes=s_t.tm_min,
+                                        seconds=s_t.tm_sec).total_seconds()
+        logging.info("Runing result: %f" % total_time)
+        return total_time
+
+    timeout = int(params.get("timeout", 360))
```
Comment (YongxueHong) [inline qemu/tests/hv_tlbflush.py:168]: Please call params.get_numeric("timeout", 360)

**Comment by kimi1978wy** [comment]
@YongxueHong  workaround the print issue now 

(1/1) Host_RHEL.m8.u3.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.hv_tlbflush.q35: PASS (721.97 s)

**Pull request merged**